index.ts

```diff
@@ -0,0 +1,425 @@
+  try {
+    const { id } = req.params;
+
+    const foundCategory = db.categories.find((category) => category.id === id);
+
+    if (!foundCategory) {
+      res.send(`Category with provided ${id} is not found`);
+      return;
+    }
+
+    const updatedCategories = db.categories.filter(
+      (category) => category.id !== id
+    );
+
+    db.categories = updatedCategories;
+
+    const updatedMaterialCategories = db.materialCategories.map(
+      (materialCategory) => {
+        if (materialCategory.categoryId === id) {
+          return { ...materialCategory, categoryId: null };
+        }
+
+        return materialCategory;
+      }
+    );
+
+    db.materialCategories = updatedMaterialCategories;
+
+    console.log("DELETE /categories/{id} DB:", db);
+
+    res.send("Category deleted successfully");
+  } catch (error) {
+    console.log(error);
+    res.send("Error in delete one category");
+  }
+});
+
+// Materials endpoints
+// Create one material
+app.post("/materials", (req: Request<{}, {}, MaterialDef, {}>, res) => {
+  try {
+    const body: unknown = req.body;
+    const parsedBody = materialDefSchema.parse(body);
+    const id = crypto.randomUUID();
+    const { categoryIds, ...materialBase } = parsedBody;
+
+    const material: Material = {
+      ...materialBase,
+      id,
+      createdAt: Date.now(),
+      updatedAt: Date.now(),
+    };
+
+    db.materials.push(material);
+
+    for (const categoryId of categoryIds) {
+      const materialCategory: MaterialCategory = { categoryId, materialId: id };
+      db.materialCategories.push(materialCategory);
+    }
+
+    console.log("POST /materials DB:", db);
+
+    res.send(material);
+  } catch (error) {
+    console.error(error);
+
+    if (error instanceof ZodError) {
+      res.status(422).send("Error creating material");
+      return;
+    }
+
+    res.status(500).send("Unknown error");
+  }
+});
+
+// Get all materials
+app.get("/materials", (req, res) => {
+  try {
+    const materials = db.materials;
+    res.send(materials);
+  } catch (error) {
+    console.log(error);
+    res.send("Error in get all materials");
+  }
+});
+
+// Get one material
+app.get("/materials/:id", (req, res) => {
+  try {
+    const { id } = req.params;
+    const foundMaterial = db.materials.find((material) => material.id === id);
+
+    if (foundMaterial) {
+      res.send(foundMaterial);
+      return;
+    }
+
+    res.send(`Material with provided ${id} is not found`);
+  } catch (error) {
+    console.log(error);
+    res.send("Error in get one material");
+  }
+});
+
+// Get all materials by one category
+app.get("/materials/category/:categoryId", (req, res) => {
+  try {
+    const { categoryId } = req.params;
+
+    const filteredMaterialIdsByCategory = db.materialCategories
+      .filter((materialCategory) => materialCategory.categoryId === categoryId)
+      .map((materialCategory) => materialCategory.materialId);
+
+    const filteredMaterials = db.materials.filter((material) =>
```

Comment:
This could be more efficient if combined with the map on line 30:

```
const filteredMaterials = db.materialCategories
  .filter((materialCategory) => materialCategory.categoryId === categoryId)
  .map((materialCategory) => db.materials.find(material => material.id === materialCategory.materialId));
```